iroh-resolver/src/unixfs_builder.rs

```diff
@@ -1,55 +1,172 @@
-use std::{fmt::Debug, path::Path, pin::Pin};
+use std::{
+    fmt::Debug,
+    path::{Path, PathBuf},
+    pin::Pin,
+};
 
 use anyhow::{anyhow, ensure, Result};
+use async_recursion::async_recursion;
 use async_trait::async_trait;
-use bytes::Bytes;
+use bytes::{Bytes, BytesMut};
 use cid::{multihash::MultihashDigest, Cid};
-use futures::{Stream, StreamExt};
+use futures::{stream::LocalBoxStream, Stream, StreamExt};
 use iroh_rpc_client::Client;
 use prost::Message;
 use tokio::io::AsyncRead;
 
 use crate::{
+    balanced_tree::TreeBuilder,
     chunker::{self, Chunker, DEFAULT_CHUNK_SIZE_LIMIT},
     codecs::Codec,
     unixfs::{dag_pb, unixfs_pb, DataType, Node, UnixfsNode},
 };
 
+// CID_LEN is the length of a cid
+const CID_LEN: usize = 64;
+
+// the PBNode.Data field may contain some data, this buffer is to ensure that
+// the contents of that data field + the size of the encoded protobuf links
+// will not be larger than the chunk size
+const BUFFER: usize = 1024;
```

Comment:
arbitrary buffer size.

https://github.com/ipfs/go-unixfs/blob/e0ef9c44e92cab793bdf321ac1531ceff418fd5f/io/directory.go#L21-L27
from `go-unixfs`, the comment on line 26 says that the "ProtoNode doesn't use the Data field so this estimate is pretty accurate)"

But as far as I can tell, in our implementation, when we encode the Protobuf Dag, we _are_ using the `data` field, so it seemed prudent to have some kind of buffer to represent the content that can exist in the field.

Reply:
@ribasushi please help